### ISSUE TITLE
Cannot output UTF8 to stdout in Python 2.6.8

### ISSUE THREAD

**Issue by me-and** [Python-Markdown/markdown#158]
In Python-Markdown 2.2.1, attempting to encode a source file containing UTF8 characters fails. Specifying the `-e` argument doesn't appear to help. Specifying `-f` to give an output file works.

This usage previously worked on Python-Markdown 2.1.1. A quick check points at commits 28110c48bd21a5ba80a1ed71e8217ce903dab8be and 1132f9e, which are attempting to resolve this sort of issue, but apparently breaks things in this circumstance.

See the below transcript:

```
$ cat test # This is a single UTF-8 en-dash followed by an LF
–

$ od -t x1 test
0000000 e2 80 93 0a
0000004

$ file test
test: UTF-8 Unicode text

$ python -m markdown.__main__ test
Traceback (most recent call last):
  File "/usr/lib/python2.6/runpy.py", line 122, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.6/runpy.py", line 34, in _run_code
    exec code in run_globals
  File "/usr/lib/python2.6/site-packages/markdown/__main__.py", line 87, in <module>
    run()
  File "/usr/lib/python2.6/site-packages/markdown/__main__.py", line 81, in run
    markdown.markdownFromFile(**options)
  File "/usr/lib/python2.6/site-packages/markdown/__init__.py", line 449, in markdownFromFile
    kwargs.get('encoding', None))
  File "/usr/lib/python2.6/site-packages/markdown/__init__.py", line 384, in convertFile
    sys.stdout.write(html)
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2013' in position 3: ordinal not in range(128)

$ python -m markdown.__main__ -e utf8 test
Traceback (most recent call last):
  File "/usr/lib/python2.6/runpy.py", line 122, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.6/runpy.py", line 34, in _run_code
    exec code in run_globals
  File "/usr/lib/python2.6/site-packages/markdown/__main__.py", line 87, in <module>
    run()
  File "/usr/lib/python2.6/site-packages/markdown/__main__.py", line 81, in run
    markdown.markdownFromFile(**options)
  File "/usr/lib/python2.6/site-packages/markdown/__init__.py", line 449, in markdownFromFile
    kwargs.get('encoding', None))
  File "/usr/lib/python2.6/site-packages/markdown/__init__.py", line 384, in convertFile
    sys.stdout.write(html)
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2013' in position 3: ordinal not in range(128)

$ python -m markdown.__main__ -f out test

$ cat out
<p>–</p>
$ python
Python 2.6.8 (unknown, Jun  9 2012, 11:30:32)
[GCC 4.5.3] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>>
```

**Comment by waylan** [comment]
Thanks for the report - and tracking down the recent relevant changes. The differences between Python 2 and 3 when working with STDOUT has become a major pain point.

If you could change [line 382](https://github.com/waylan/Python-Markdown/blob/28110c48bd21a5ba80a1ed71e8217ce903dab8be/markdown/__init__.py#L382) of markdown/**init**.py to read `if sys.stdout.encoding == encoding:` does that solve your problem?

**Comment by me-and** [comment]
Yes, that change seems to have everything working.

For the sake of being absolutely explicit, the diff:

```
diff --git a/markdown/__init__.py b/markdown/__init__.py
index cda6b76..d7640a4 100644
--- a/markdown/__init__.py
+++ b/markdown/__init__.py
@@ -379,7 +379,7 @@ class Markdown:
                 output_file.write(html)
                 # Don't close here. User may want to write more.
         else:
-            if sys.stdout.encoding:
+            if sys.stdout.encoding == encoding:
                 # If we are in Python 3 or if we are not piping output:
                 sys.stdout.write(html)
             else:
```

**Comment by waylan** [comment]
Excellent. I'll have to run some tests in Python 3 tonight to make sure that doesn't break anything.

**Comment by me-and** [comment]
Wonderful. Thanks for the prompt resolution!

**Comment by me-and** [comment]
…I say that. I've just pulled the tip, and now see the following:

```
$ python -m markdown.__main__  test
Traceback (most recent call last):
  File "/usr/lib/python2.6/runpy.py", line 122, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.6/runpy.py", line 34, in _run_code
    exec code in run_globals
  File "/home/add/Python-Markdown/markdown/__main__.py", line 87, in <module>
    run()
  File "/home/add/Python-Markdown/markdown/__main__.py", line 81, in run
    markdown.markdownFromFile(**options)
  File "/home/add/Python-Markdown/markdown/__init__.py", line 448, in markdownFromFile
    kwargs.get('encoding', None))
  File "/home/add/Python-Markdown/markdown/__init__.py", line 383, in convertFile
    html = html.encode(encoding, errors="xmlcharrefreplace")
TypeError: encode() takes no keyword arguments
```

According to http://docs.python.org/2.7/library/stdtypes.html#str.encode, the keyword argument was added in Python 2.7. The following change fixes this for Python 2.6.8. Let me know if you'd rather this as a pull request and/or in a new issue.

```
diff --git a/markdown/__init__.py b/markdown/__init__.py
index 1b71295..75c179b 100644
--- a/markdown/__init__.py
+++ b/markdown/__init__.py
@@ -380,7 +380,7 @@ class Markdown:
                 # Don't close here. User may want to write more.
         else:
             # Encode manually and write bytes to stdout.
-            html = html.encode(encoding, errors="xmlcharrefreplace")
+            html = html.encode(encoding, "xmlcharrefreplace")
             try:
                 # Write bytes directly to buffer (Python 3).
                 sys.stdout.buffer.write(html)
```

(I'm not sure how GitHub's issue tracker works, so I'm not sure if you'll get this now the ticket is closed. If I don't get a response in a couple of days, I'll open a new issue or create a pull request.)
